File: test/aml_comparisons/gurobi/models/misc/nqueens.cpp
//#define _GLIBCXX_USE_CXX11_ABI 0
#include <map>
#include <vector>
#include "gurobi_c++.h"


void nqueens_scalar(GRBModel& model, size_t N)
{
std::vector<std::vector<GRBVar>> x(N, std::vector<GRBVar>(N));
for (size_t i=0; i<N; i++)
    for (size_t j=0; j<N; j++)
        x[i][j] = model.addVar(0,1,0,GRB_BINARY);

// obj
GRBLinExpr obj;
for (size_t i=0; i<N; i++)
    for (size_t j=0; j<N; j++)
        obj += x[i][j];
model.setObjective( obj );

// one per row
for (size_t i=0; i<N; i++) {
    GRBLinExpr c;
    for (size_t j=0; j<N; j++)
        c += x[i][j];
    model.addConstr( c == 1 );
    }

// one per column
for (size_t j=0; j<N; j++) {
    GRBLinExpr c;
    for (size_t i=0; i<N; i++)
        c += x[i][j];
    model.addConstr( c == 1 );
    }

// \diagonals_col
for (size_t i=0; i<N-1; i++) {
    GRBLinExpr c;
    c += x[0][i];
    for (size_t j=1; j<N-i; j++)
        c += x[j][i+j];
    model.addConstr( c <= 1 );
    }
// \diagonals_row
for (size_t i=1; i<N-1; i++) {
    GRBLinExpr c;
    c += x[i][0];
    for (size_t j=1; j<N-i; j++)
        c += x[i+j][j];
    model.addConstr( c <= 1 );
    }

// /diagonals_col
for (size_t i=1; i<N; i++) {
    GRBLinExpr c;
    c += x[0][i];
    for (size_t j=1; j<=i; j++)
        c += x[j][i-j];
    model.addConstr( c <= 1 );
    }
// /diagonals_row
for (size_t i=1; i<N-1; i++) {
    GRBLinExpr c;
    c += x[i][N-1];
    for (size_t j=1; j<N-i; j++)
        c += x[i+j][N-1-j];
    model.addConstr( c <= 1 );
    }
}

File: test/aml_comparisons/pygurobi/models/nqueens.py
import sys
from gurobipy import Model, quicksum


def nqueens(N):
    model = Model("nqueens")

    x = model.addVars(N, N, lb=0.0, ub=1.0, vtype='C')

    # obj
    model.setObjective(x.sum())

    # one per row
    model.addConstrs((x.sum(i, '*') == 1 for i in range(N)))

    # one per column
    model.addConstrs((x.sum('*', j) == 1 for j in range(N)))

    # \diagonals_col
    model.addConstrs((x[0, i] + quicksum(x[j, i + j]
                                         for j in range(1, N - i)) <= 1
                      for i in range(N - 1)))
    # \diagonals_row
    model.addConstrs((x[i, 0] + quicksum(x[i + j, j]
                                         for j in range(1, N - i)) <= 1
                      for i in range(1, N - 1)))
    # /diagonals_col
    model.addConstrs((x[0, i] + quicksum(x[j, i - j]
                                         for j in range(1, i + 1)) <= 1
                      for i in range(1, N)))
    # /diagonals_row
    model.addConstrs((x[i, N - 1] + quicksum(x[i + j, N - 1 - j]
                                             for j in range(1, N - i)) <= 1
                      for i in range(1, N)))

    return model

